vpr/src/place/place.cpp

```diff
@@ -2611,13 +3103,88 @@ static void get_non_updateable_bb(ClusterNetId net_id, t_bb* bb_coord_new) {
      * clip to 1 in both directions as well (since minimum channel index *
      * is 0).  See route_common.cpp for a channel diagram.               */
 
-    bb_coord_new->xmin = max(min<int>(xmin, device_ctx.grid.width() - 2), 1);  //-2 for no perim channels
-    bb_coord_new->ymin = max(min<int>(ymin, device_ctx.grid.height() - 2), 1); //-2 for no perim channels
-    bb_coord_new->xmax = max(min<int>(xmax, device_ctx.grid.width() - 2), 1);  //-2 for no perim channels
-    bb_coord_new->ymax = max(min<int>(ymax, device_ctx.grid.height() - 2), 1); //-2 for no perim channels
+    bb_coord_new.xmin = max(min<int>(xmin, device_ctx.grid.width() - 2), 1);  //-2 for no perim channels
+    bb_coord_new.ymin = max(min<int>(ymin, device_ctx.grid.height() - 2), 1); //-2 for no perim channels
+    bb_coord_new.xmax = max(min<int>(xmax, device_ctx.grid.width() - 2), 1);  //-2 for no perim channels
+    bb_coord_new.ymax = max(min<int>(ymax, device_ctx.grid.height() - 2), 1); //-2 for no perim channels
 }
 
-static void update_bb(ClusterNetId net_id, t_bb* bb_coord_new, t_bb* bb_edge_new, int xold, int yold, int xnew, int ynew) {
+static void get_non_updateable_layer_bb(ClusterNetId net_id,
+                                        std::vector<t_2D_bb>& bb_coord_new,
+                                        std::vector<int>& num_sink_layer) {
+    //TODO: account for multiple physical pin instances per logical pin
+
+    auto& device_ctx = g_vpr_ctx.device();
+    int num_layers = device_ctx.grid.get_num_layers();
+    num_sink_layer = std::vector<int>(num_layers, 0);
+    std::vector<int> xmin(num_layers, OPEN);
```

Comment:
Why the xmin, ymin temporaries again?  Unnecessary allocations, deletes and copying.

Reply:
The reason for creating `xmin`, `ymin`, etc. is because I need to include the source block in all layers, so I initialize them with the source block coordinates. Later, if there is no sink on a particular layer, I want to assign `OPEN` to the bounding box coordinates of that layer. Additionally, at the end of the function, I perform additional checks to ensure the bounding box coordinates are inside the grid. These reasons led me to create temporary values to obtain the bounding box coordinates on each layer and store the valid coordinates in the main data structure.

Reply:
You will pay a runtime price for this, so if it's a key loop it's worth thinking about whether you need the vector temporaries or could recode without them.

Reply:
I will attempt to rewrite it without utilizing these temporary vectors. Although I initially chose this approach for its cleanliness, I agree that given the inner loop context, it is worth experimenting with alternatives.